actionpack/lib/action_dispatch/journey/path/pattern.rb

```diff
@@ -27,21 +28,12 @@ def eager_load!
           required_names
           offsets
           to_regexp
-          nil
-        end
-
-        def ast
-          @spec.find_all(&:symbol?).each do |node|
-            re = @requirements[node.to_sym]
-            node.regexp = re if re
-          end
-
-          @spec
+          @ast = nil
```

Comment:
Do we need to set the `@ast` to `nil`? We have code that get two patterns and join then and I can't find a way to made that code work if we don't have the original ast anymore.

Reply:
Not to say that this breaks other methods in this object, so the object will be in an invalid state after eager load.

Reply:
If the reason to not keep the `@ast` is to avoid to keep the arrays that are stored in the `Ast` object, we could keep the `ast.tree` like it was being done before.

```diff
diff --git a/actionpack/lib/action_dispatch/journey/path/pattern.rb b/actionpack/lib/action_dispatch/journey/path/pattern.rb
index 4cd862f791..41a04fdcc0 100644
--- a/actionpack/lib/action_dispatch/journey/path/pattern.rb
+++ b/actionpack/lib/action_dispatch/journey/path/pattern.rb
@@ -4,11 +4,11 @@ module ActionDispatch
   module Journey # :nodoc:
     module Path # :nodoc:
       class Pattern # :nodoc:
-        attr_reader :ast, :names, :requirements, :anchored
-        alias :spec :ast
+        attr_reader :ast, :names, :requirements, :anchored, :spec
 
         def initialize(ast, requirements, separators, anchored)
           @ast          = ast
+          @spec         = ast.tree
           @requirements = requirements
           @separators   = separators
           @anchored     = anchored
```

@composerinteralia what do you think?

Reply:
I added the line to clear out the `Ast` objects after eager load in response to concerns about increasing retained memory. I was aware this would break some methods, but those methods are not public api and not called after eager load so I thought it would be OK.

Bringing back the original behavior of `spec` to return the root `Node` rather than an `Ast` makes sense to me. That way we can clear out the `Ast` and still be able to return the root node. I've opened https://github.com/rails/rails/pull/43006 to do that, as well as clean up a few other small details.

I'm not sure we need to cache `requirements_anchored`, unless there's a use case for calling that after eager_load. We use it to partition the route, which only happens when first adding the route.